dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/WebAppSecurityCollaboratorImpl.java

```diff
@@ -461,31 +461,16 @@ public void handleException(HttpServletRequest req, HttpServletResponse rsp, Thr
     }
 
     @Override
-    @FFDCIgnore(NumberFormatException.class)
     public boolean isCDINeeded() {
-        int specLevel = WebContainer.getServletContainerSpecLevel();
-
-        /*
-         * Determine which version of appSecurity we are running.
-         */
-        Set<String> installedFeatures = provisionerService.getInstalledFeatures();
-        float appSecurityVers = 0.0f;
-        for (String feature : installedFeatures) {
-            if (feature.startsWith("appSecurity-")) {
-                try {
-                    appSecurityVers = Float.valueOf(feature.substring(12));
-                    break;
-                } catch (NumberFormatException e) {
-                    // Ignore. Not the feature we are looking for.
-                }
-            }
-        }
-
         /*
-         * Future-proof check. Some of these combinations aren't valid, but
-         * they will not make it this far.
+         * Tried to future-proof this check by iterating over all the
+         * installed features and finding the version of appSecurity installed,
+         * but that introduced a performance degradation. So for now, checking
+         * for the appSecurity-3.0/4.0 features directly.
          */
-        return appSecurityVers >= 3.0 && specLevel >= WebContainer.SPEC_LEVEL_40;
+        return WebContainer.getServletContainerSpecLevel() >= WebContainer.SPEC_LEVEL_40 &&
```

Comment:
In general this whole method is an anti-pattern.  Usually this is handled with a service being applied to the instance and this not being calculated on EVERY web request.  In order to reduce the extra processing of getting the installed features I suggest to change this to the below for now in order to get us back to close to where we were.

```
        if (WebContainer.getServletContainerSpecLevel() < WebContainer.SPEC_LEVEL_40) {
            return false;
        }
        Set<String> features = provisionerService.getInstalledFeatures();
        return features.contains("appSecurity-3.0") || features.contains("appSecurity-4.0");
```

Reply:
Made this change.